### PR TITLE
V1/channels

### PR DESCRIPTION
**Pull request by alecritson**
This brings Channels into v1. There are some additions to Actions which may need to be carried across to Addresses, but that should be minimal overhead.

### Remove `Action` suffix from the Action class names.

This was inspired by this [tweet thread](https://twitter.com/jeffrey_way/status/1297944344603824129) I think not having the suffix doesn't add any issues with the description of the class and what's happening. I think it also makes the syntax a little nicer. For example.

Instead of

```
FetchChannelAction::run();
```

It's just

```
FetchChannel::run();
```

I don't think it'll ever conflict? I'm all for Suffixing classes when it's helpful and makes sense, but I think for actions we probably won't need to worry about it? We know we're using actions, so we can assume when we see something like that, it's an action. If we needed a job we would suffix that class like `FetchChannelJob` as that's descriptive and makes sense in the context it's being run.

Again. Open to pushback on this, just a thought, whichever way we go it's easy to adjust, least then we know going forward.

### Added `AbstractAction` class with some helpers

When returning relations for resources we use an `include` query param. To get a relations count we can use the `counts` query param (open to change ofc)

Now actions extend this abstract class which have the helper methods:

- `$this->resolveEagerRelations()` - Will automatically parse the `include` attribute on an action into an array from a comma separated string
- `$this->resolveRelationCounts()` - Does the same as above, but for when including counts.

### Added `GetCandy\Api\Core\Traits\ReturnsJsonResponses` trait to relevant actions

Some actions return json responses, this trait makes it easy to keep those responses consistent with helper methods.

### Docs

Draft docs are here https://candy-api.readme.io/docs/channels (visible to team only)

### CODE REVIEW

File: routes/api.client.php [hunk 1]
```diff
@@ -17,13 +17,12 @@
     return response()->json([
         'version' => GetCandy::version(),
         'locale' => app()->getLocale(),
-        'channel' => new GetCandy\Api\Http\Resources\Channels\ChannelResource($channel->getChannel()),
+        'channel' => new \GetCandy\Api\Core\Channels\Resources\ChannelResource($channel->getChannel()),
```
Comment (glennjacobs) [inline routes/api.client.php:20]: We seem to be inconsistent with the backslashes at the start of namespaces, e.g. some are

`new GetCandy`

others

`new \GetCandy`

Can we keep to one?

File: src/Core/Traits/HasAttributes.php [hunk 2]
```diff
@@ -45,7 +47,7 @@ public function attribute($handle, $channel = null, $locale = null)
         }
 
         if (empty($this->attribute_data[$handle][$channel])) {
-            $defaultChannel = GetCandy::channels()->getDefaultRecord();
+            $defaultChannel = FetchDefaultChannel::run();
```
Comment (glennjacobs) [inline src/Core/Traits/HasAttributes.php:50]: `FetchDefaultChannel` is that the right name? Or should it be more along the lines of `FetchSelectedChannel`? I'm not too sure what this does haha

**Review by glennjacobs — CHANGES_REQUESTED** [review]

File: src/Core/Traits/HasAttributes.php (hunk 2 above)
Comment (glennjacobs) [reply src/Core/Traits/HasAttributes.php:50]: Oh and we talked about the Channel closure idea.

**Comment by glennjacobs** [comment]
I'm happy with the Action naming convention. I think suffixes are helpful when there are different types of classes in a folder. But if they're in an "Actions" folder, it's pretty obvious what's going on.

We talked about the Channel selection closure idea.

Documentation looks pretty good. Few grammar issues, but we can hit those later.

File: src/Core/Traits/HasAttributes.php (hunk 2 above)
Comment (alecritson) [reply src/Core/Traits/HasAttributes.php:50]: This doesn't send back the current channel, just the one that's been marked as `default` in the database.

Comment (glennjacobs) [reply src/Core/Traits/HasAttributes.php:50]: OK

File: routes/api.client.php (hunk 1 above)
Comment (alecritson) [reply routes/api.client.php]: I've changed this now to use an action instead. Was thinking of even moving this logic out into the GetCandy facade, but not really part of this development...

**Review by glennjacobs — APPROVED** [review]

**Pull request merged**